File: C++/2-Medium/00230._二叉搜索树中第K小的元素/00230._二叉搜索树中第K小的元素.cpp
#include <iostream>
#include <vector>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

// Helper function to create a binary tree from a vector of integers (null represented by -1)
TreeNode* createTree(const vector<int>& nodes, int index) {
    if (index >= nodes.size() || nodes[index] == -1) return nullptr;
    
    TreeNode* root = new TreeNode(nodes[index]);
    root->left = createTree(nodes, 2 * index + 1);
    root->right = createTree(nodes, 2 * index + 2);
    return root;
}

// Function to perform in-order traversal
void inOrderTraversal(TreeNode* node, int &k, int &result) {
    if (node == NULL || k == 0) return;

    // Traverse the left subtree
    inOrderTraversal(node->left, k, result);

    // Visit the node
    if (--k == 0) {
        result = node->val;
        return;
    }

    // Traverse the right subtree
    inOrderTraversal(node->right, k, result);
}

// Main function to find the kth smallest element in a BST
int kthSmallest(TreeNode* root, int k) {
    int result = -1; // Initialize result
    inOrderTraversal(root, k, result);
    return result;
}

int main() {
    // Example: create a binary search tree
    vector<int> nodes = {5, 3, 6, 2, 4, -1, -1, 1}; // BST representation (level-order)
    TreeNode* root = createTree(nodes, 0);
    
    int k = 3; // Find the 3rd smallest element
    cout << "The " << k << "rd smallest element in the BST is: " << kthSmallest(root, k) << endl;
    
    return 0;
}

File: Python/2-Medium/00230._二叉搜索树中第K小的元素/00230._二叉搜索树中第K小的元素.py
from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        """
        在二叉搜索树中查找第K小的元素。
        
        Args:
        - root: TreeNode, 二叉搜索树的根节点。
        - k: int, 指定的第K小的元素的序号。
        
        Returns:
        - int, 第K小的元素的值。
        
        示例:
        >>> tree = TreeNode(3, TreeNode(1, None, TreeNode(2)), TreeNode(4))
        >>> sol = Solution()
        >>> sol.kthSmallest(tree, 1)
        1
        
        """
        # 使用中序遍历的方法，初始化计数器和结果变量
        self.k = k
        self.result = None
        self.count = 0
        
        def inorder(node: Optional[TreeNode]):
            """ 中序遍历辅助函数 """
            if not node or self.result is not None:
                return
            
            inorder(node.left)
            self.count += 1
            if self.count == self.k:
                self.result = node.val
                return
            inorder(node.right)
        
        inorder(root)
        return self.result

# 构建示例二叉树进行测试
tree = TreeNode(3, TreeNode(1, None, TreeNode(2)), TreeNode(4))
sol = Solution()
# 查找第1小的元素
result = sol.kthSmallest(tree, 1)
print(result)  # 预期输出：1